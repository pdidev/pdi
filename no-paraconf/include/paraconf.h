/*******************************************************************************
* Copyright (C) 2025 Commissariat a l'energie atomique et aux energies alternatives (CEA)
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are met:
* * Redistributions of source code must retain the above copyright
*   notice, this list of conditions and the following disclaimer.
* * Redistributions in binary form must reproduce the above copyright
*   notice, this list of conditions and the following disclaimer in the
*   documentation and/or other materials provided with the distribution.
* * Neither the name of CEA nor the names of its contributors may be used to
*   endorse or promote products derived from this software without specific
*   prior written permission.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
******************************************************************************/

/** \file paraconf.h for deactivation
 */

#ifndef PARACONF_H_
#define PARACONF_H__H_

#ifdef __cplusplus
extern "C" {
#endif

/** \addtogroup error
 * \{
 */

typedef struct PC_tree_s {
} PC_tree_t;
struct _IO_FILE;
typedef struct _IO_FILE FILE;

static inline PC_tree_t PC_parse_string(const char* document)
{
	PC_tree_t mock = {};
	return mock;
}

static inline PC_tree_t PC_parse_file(FILE *file)
{
	PC_tree_t mock = {};
	return mock;
}

#ifdef __cplusplus
} // extern C
#endif

#endif // PARACONF_H_
