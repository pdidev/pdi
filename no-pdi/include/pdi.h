/*******************************************************************************
* Copyright (C) 2025 Commissariat a l'energie atomique et aux energies alternatives (CEA)
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are met:
* * Redistributions of source code must retain the above copyright
*   notice, this list of conditions and the following disclaimer.
* * Redistributions in binary form must reproduce the above copyright
*   notice, this list of conditions and the following disclaimer in the
*   documentation and/or other materials provided with the distribution.
* * Neither the name of CEA nor the names of its contributors may be used to
*   endorse or promote products derived from this software without specific
*   prior written permission.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
******************************************************************************/

/** \file pdi.h for deactivation
 *
 * C user API
 *
 * The user facing API is the interface offered by PDI to C application
 * developers.
 *
 * \defgroup init_final Initialization and finalization
 *
 * The initialization and finalization part of the API is used to setup PDI,
 * release its resources and check version information.
 *
 * \defgroup annotation Code annotation
 *
 * The code annotation API is the main interface to use in the code.
 *
 * It offers functions that can be called from code with no side effect by
 * default and that can therefore be considered as annotations.
 *
 * \defgroup error Error handling
 *
 * The error handling API supports checking the error status of PDI.
 *
 * By default, errors in PDI C API are signaled by a return code of type
 * PDI_status_t and an error message can be retrieved with the PDI_errmsg
 * function. This default behavior can be changed by replacing the error handler
 * with the PDI_errhandler function.
 *
 */

#ifndef PDI_H_
#define PDI_H_

#ifdef __cplusplus
extern "C" {
#endif

/** \addtogroup error
 * \{
 */

/** Error codes of PDI
 */
typedef enum PDI_status_e {
	/// everything went well
	PDI_OK = 0,
	/// on an input call, no such data is available
	PDI_UNAVAILABLE,
	/// The configuration file is invalid
	PDI_ERR_CONFIG,
	/// A value expression is invalid
	PDI_ERR_VALUE,
	/// Tried to load a non-existing plugin
	PDI_ERR_PLUGIN,
	/// Implementation limitation (typically an unimplemented feature)
	PDI_ERR_IMPL,
	/// A system error occured (OS, etc.)
	PDI_ERR_SYSTEM,
	/** A call to a function has been made at a wrong time (e.g. closing an
	 *  unopened transaction)
	 */
	PDI_ERR_STATE,
	/// A conflict of onwership over a content has been raised
	PDI_ERR_RIGHT,
	/// Invalid type error
	PDI_ERR_TYPE

} PDI_status_t;

/** Type of a callback function used when an error occurs
 * \param status the error code
 * \param message the human-readable error message
 * \param context a user-provided context
 */
typedef void (*PDI_errfunc_f)(PDI_status_t status, const char* message, void* context);

/** Definition of an error handler
 */
typedef struct PDI_errhandler_s {
	/// The function to handle the error (none if NULL)
	PDI_errfunc_f func;

	/// the context that will be provided to the function
	void* context;

} PDI_errhandler_t;

#ifndef PDI_NULL_HANDLER_DEFINED
#define PDI_NULL_HANDLER_DEFINED
/** Does nothing
 */
const PDI_errhandler_t PDI_NULL_HANDLER = {NULL, NULL};

/** Prints the error message and aborts if the status is invalid
 */
const PDI_errhandler_t PDI_ASSERT_HANDLER = PDI_NULL_HANDLER;

/** Prints the error message and continue if the status is invalid
 */
const PDI_errhandler_t PDI_WARN_HANDLER = PDI_NULL_HANDLER;
#endif

/** Return a human-readabe message describing the last error that occured in PDI
 */
static inline const char* PDI_errmsg(void)
{
	return "";
}

/** Sets the error handler to use
 */
static inline PDI_errhandler_t PDI_errhandler(PDI_errhandler_t)
{
	return PDI_NULL_HANDLER;
}

/// \}

/** \addtogroup init_final Initialization and finalization
 *
 * The initialization and finalization part of the API is used to setup PDI,
 * release its resources and check version information.
 * \{
 */

/** Initializes PDI
 */
static inline PDI_status_t PDI_init(PC_tree_t)
{
	return PDI_OK;
}

/** Finalizes PDI
 * \return an error status
 */
static inline PDI_status_t PDI_finalize(void)
{
	return PDI_OK;
}

/** Checks PDI API version
 */
static inline PDI_status_t PDI_version(unsigned long*, unsigned long)
{
	return PDI_OK;
}

/// \}

/** \addtogroup annotation
 * \{
 */

/**
 * Access directions
 */
typedef enum PDI_inout_e {
	/// No data transfert
	PDI_NONE = 0,
	/// data tranfer from PDI to the main code
	PDI_IN = 1,
	/// data transfer from the main code to PDI
	PDI_OUT = 2,
	/// data transfer in both direction
	PDI_INOUT = 3

} PDI_inout_t;

/** Shares some data with PDI. The user code should not modify it before
 * a call to either PDI_release or PDI_reclaim.
 */
static inline PDI_status_t PDI_share(const char*, void*, PDI_inout_t)
{
	return PDI_OK;
}

/** Requests for PDI to access a data buffer.
 */
static inline PDI_status_t PDI_access(const char*, void**, PDI_inout_t)
{
	return PDI_OK;
}

/** Releases ownership of a data shared with PDI. PDI is then responsible to
 * free the associated memory whenever necessary.
 */
static inline PDI_status_t PDI_release(const char*)
{
	return PDI_OK;
}

/** Reclaims ownership of a data buffer shared with PDI. PDI does not manage
 * the buffer memory anymore.
 */
static inline PDI_status_t PDI_reclaim(const char*)
{
	return PDI_OK;
}

/** Triggers a PDI "event"
 */
static inline PDI_status_t PDI_event(const char*)
{
	return PDI_OK;
}

/** Shortly exposes some data to PDI. Equivalent to PDI_share + PDI_reclaim.
 */
static inline PDI_status_t PDI_expose(const char*, void*, PDI_inout_t)
{
	return PDI_OK;
}

/** Performs multiple exposes at once. All the data is shared in order they were specified
 *  and reclaimed in reversed order after an event is triggered.
 *
 *  NULL argument indicates an end of the list.
 */
static inline PDI_status_t PDI_multi_expose(const char*, const char*, void*, PDI_inout_t, ...)
{
	return PDI_OK;
}

#ifdef PDI_WITH_DEPRECATED

/** Begin a transaction in which all PDI_expose calls are grouped.
 *
 * This requires a call to PDI_transaction_end to close the transaction.
 *
 * \deprecated the transaction part of the API is deprecated, the
 * PDI_multi_expose function should be used instead.
 *
 * \see PDI_expose the function used to expose data inside the transaction
 * \see PDI_transaction_end the function used to end the transaction
 *
 * \return an error status
 */
static inline PDI_status_t PDI_transaction_begin(const char*)
{
	return PDI_OK;
}

/** Ends the previously opened transaction.
 *
 * \deprecated the transaction part of the API is deprecated, the
 * PDI_multi_expose function should be used instead.
 *
 * \see PDI_transaction_begin the function used to start the transaction
 * \see PDI_expose the function used to expose data inside the transaction
 *
 * \return an error status
 */
static inline PDI_status_t PDI_transaction_end(void)
{
	return PDI_OK;
}

#endif // PDI_WITH_DEPRECATED

/// \}

#ifdef __cplusplus
} // extern C
#endif


#endif // PDI_H_
