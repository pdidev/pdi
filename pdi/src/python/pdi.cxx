/*******************************************************************************
* Copyright (C) 2015-2024 Commissariat a l'energie atomique et aux energies alternatives (CEA)
* Copyright (C) 2020 Institute of Bioorganic Chemistry Polish Academy of Science (PSNC)
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are met:
* * Redistributions of source code must retain the above copyright
*   notice, this list of conditions and the following disclaimer.
* * Redistributions in binary form must reproduce the above copyright
*   notice, this list of conditions and the following disclaimer in the
*   documentation and/or other materials provided with the distribution.
* * Neither the name of CEA nor the names of its contributors may be used to
*   endorse or promote products derived from this software without specific
*   prior written permission.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
******************************************************************************/

#include "config.h"

#include <cstddef>
#include <exception>
#include <iostream>
#include <sstream>
#include <string>
#include <type_traits>
#include <unordered_set>

#include <pybind11/numpy.h>
#include <pybind11/pybind11.h>

#include "pdi/context.h"
#include "pdi/data_descriptor.h"
#include "pdi/error.h"
#include "pdi/paraconf_wrapper.h"
#include "pdi/plugin.h"
#include "pdi/python/python_ref_wrapper.h"
#include "pdi/python/tools.h"
#include "pdi/ref_any.h"

#include "global_context.h"

namespace {

using namespace PDI;
// template <typename T> using pycls = pybind11::class_<T>;
template <typename T>
using pyenu = pybind11::enum_<T>;
using pyarr = pybind11::array;
// using pylst = pybind11::list;
using pymod = pybind11::module;
using pyobj = pybind11::object;
// using pystr = pybind11::str;
using pytup = pybind11::tuple;
using namespace pybind11::literals;
using std::move;
using std::vector;

/** Check API version of PDI and python
 *
 * \param pyexpected tuple with expected PDI version
 * \return python tulpe with PDI version
 */
pytup pyversion(pytup pyexpected)
{
	Paraconf_wrapper fw;
	
	unsigned long expected_major = pyexpected[0].cast<unsigned long>();
	unsigned long expected_minor = pyexpected[1].cast<unsigned long>();
	unsigned long expected_patch = pyexpected[2].cast<unsigned long>();
	
	if ((expected_major || expected_minor || expected_patch) && (expected_major != PDI_VERSION_MAJOR || expected_minor > PDI_VERSION_MINOR)) {
		throw Plugin_error{
			"Invalid PDI API version: %lu.%lu.%lu, PDI provided version is %lu.%lu.%lu",
			expected_major,
			expected_minor,
			expected_patch,
			PDI_VERSION_MAJOR,
			PDI_VERSION_MINOR,
			PDI_VERSION_PATCH};
	}
	
	return pybind11::make_tuple(PDI_VERSION_MAJOR, PDI_VERSION_MINOR, PDI_VERSION_PATCH);
}

} // namespace

/** Macro that creates entry point in python interpreter
 * \param _pdi name of the module
 * \param m macro argument that defines a variable of type py::module
 */
PYBIND11_MODULE(_pdi, m)
{
	m.attr("__name__") = "pdi._pdi"; // make this a private submodule of pdi
	m.doc() = "PDI python public application API";
	
	pybind11::register_exception<Error>(m, "Error");
	
	pyenu<PDI_inout_t>(m, "Inout").value("IN", PDI_IN).value("OUT", PDI_OUT).value("INOUT", PDI_INOUT).value("NONE", PDI_NONE).export_values();
	
	m.def("version", pyversion, "Checks PDI API version", "version"_a = pybind11::make_tuple(0, 0, 0));
	
	m.def(
	    "init",
	[](char* conf) {
		Paraconf_wrapper fw;
		Global_context::init(PC_parse_string(conf));
	},
	"Initialize PDI"
	    );
	    
	m.def(
	    "finalize",
	[]() {
		Paraconf_wrapper fw;
		Global_context::finalize();
	},
	"Finalize PDI"
	    );
	    
	m.def(
	    "event",
	[](const char* name) {
		Paraconf_wrapper fw;
		Global_context::context().event(name);
	},
	"Triggers a PDI \"event\""
	    );
	    
	m.def(
	    "share",
	[](const char* name, pybind11::array pybuf, PDI_inout_t access) {
		Paraconf_wrapper fw;
		Ref r{
			pybuf.mutable_data(),
			[pybuf](void*) { /* keep pybuf copy to prevent deallocation of the underlying memory */ },
			python_type(pybuf),
			static_cast<bool>(access& PDI_OUT),
			static_cast<bool>(access& PDI_IN)};
		try {
			Global_context::context()[name].share(r, false, false);
		} catch (...) {
			// on error, do not free the data as would be done automatically otherwise
			r.release();
			throw;
		}
	},
	"Shares some data with PDI. The user code should not modify it before a call to either release or reclaim"
	    );
	    
	m.def(
	    "access",
	[](const char* name, PDI_inout_t inout) {
		Paraconf_wrapper fw;
		Data_descriptor& desc = Global_context::context()[name];
		desc.share(desc.ref(), false, false);
		return to_python(desc.ref(), !(inout& PDI_OUT));
	},
	"Requests for PDI to access a data buffer"
	    );
	    
	m.def(
	    "release",
	[](const char* name) {
		Paraconf_wrapper fw;
		Global_context::context()[name].release();
	},
	"Releases ownership of a data shared with PDI. PDI is then responsible to free the associated memory whenever necessary."
	    );
	    
	m.def(
	    "reclaim",
	[](const char* name) {
		Paraconf_wrapper fw;
		Global_context::context()[name].reclaim();
	},
	"Reclaims ownership of a data buffer shared with PDI. PDI does not manage the buffer memory anymore."
	    );
	    
	pybind11::class_<Python_ref_wrapper>(m, "ref")
	.def("__getattribute__", &Python_ref_wrapper::getattribute) // get member
	.def("__setattr__", &Python_ref_wrapper::setattribute) // set member
	.def("__getitem__", &Python_ref_wrapper::getitem); // [] operator
}
