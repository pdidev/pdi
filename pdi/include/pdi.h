/*******************************************************************************
* Copyright (C) 2015-2021 Commissariat a l'energie atomique et aux energies alternatives (CEA)
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are met:
* * Redistributions of source code must retain the above copyright
*   notice, this list of conditions and the following disclaimer.
* * Redistributions in binary form must reproduce the above copyright
*   notice, this list of conditions and the following disclaimer in the
*   documentation and/or other materials provided with the distribution.
* * Neither the name of CEA nor the names of its contributors may be used to
*   endorse or promote products derived from this software without specific
*   prior written permission.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
******************************************************************************/

/** \file pdi.h
 *
 * C user API
 *
 * The user facing API is the interface offered by PDI to C application
 * developers.
 *
 * \defgroup init_final Initialization and finalization
 *
 * The initialization and finalization part of the API is used to setup PDI,
 * release its resources and check version information.
 *
 * \defgroup annotation Code annotation
 *
 * The code annotation API is the main interface to use in the code.
 *
 * It offers functions that can be called from code with no side effect by
 * default and that can therefore be considered as annotations.
 *
 * \defgroup error Error handling
 *
 * The error handling API supports checking the error status of PDI.
 *
 * By default, errors in PDI C API are signaled by a return code of type
 * PDI_status_t and an error message can be retrieved with the PDI_errmsg
 * function. This default behavior can be changed by replacing the error handler
 * with the PDI_errhandler function.
 * 
 */

#ifndef PDI_H_
#define PDI_H_

#include <paraconf.h>

#include <pdi/export.h>
#include <pdi/version.h>

//#include "/gpfs/softs/spack_0.17/opt/spack/linux-centos7-cascadelake/gcc-11.2.0/openmpi-4.1.1-ujlwnrlh5sewm2rxkpio3h5mariwgetn/include/mpi.h"

#ifdef __cplusplus
extern "C" {
#endif

/** \addtogroup error
 * \{
 */

/** Error codes of PDI
 */
typedef enum PDI_status_e {
	/// everything went well
	PDI_OK = 0,
	/// on an input call, no such data is available
	PDI_UNAVAILABLE,
	/// The configuration file is invalid
	PDI_ERR_CONFIG,
	/// A value expression is invalid
	PDI_ERR_VALUE,
	/// Tried to load a non-existing plugin
	PDI_ERR_PLUGIN,
	/// Implementation limitation (typically an unimplemented feature)
	PDI_ERR_IMPL,
	/// A system error occured (OS, etc.)
	PDI_ERR_SYSTEM,
	/** A call to a function has been made at a wrong time (e.g. closing an
	 *  unopened transaction)
	 */
	PDI_ERR_STATE,
	/// A conflict of onwership over a content has been raised
	PDI_ERR_RIGHT,
	/// Invalid type error
	PDI_ERR_TYPE

} PDI_status_t;

/** Type of a callback function used when an error occurs
 * \param status the error code
 * \param message the human-readable error message
 * \param context a user-provided context
 */
typedef void (*PDI_errfunc_f)(PDI_status_t status, const char* message, void* context);

/** Definition of an error handler
 */
typedef struct PDI_errhandler_s {
	/// The function to handle the error (none if NULL)
	PDI_errfunc_f func;

	/// the context that will be provided to the function
	void* context;

} PDI_errhandler_t;

/** Prints the error message and aborts if the status is invalid
 */
extern const PDI_errhandler_t PDI_EXPORT PDI_ASSERT_HANDLER;

/** Prints the error message and continue if the status is invalid
 */
extern const PDI_errhandler_t PDI_EXPORT PDI_WARN_HANDLER;

/** Does nothing
 */
extern const PDI_errhandler_t PDI_EXPORT PDI_NULL_HANDLER;


/** Return a human-readabe message describing the last error that occured in PDI
 */
const char PDI_EXPORT* PDI_errmsg(void);

/** Sets the error handler to use
 *
 * PDI_asserthandler is the default handler before this function is called
 *
 * \param handler the new handler to set
 * \return the previous handler
 */
PDI_errhandler_t PDI_EXPORT PDI_errhandler(PDI_errhandler_t handler);

/// \}

/** \addtogroup init_final Initialization and finalization
 *
 * The initialization and finalization part of the API is used to setup PDI,
 * release its resources and check version information.
 * \{
 */

/** Initializes PDI
 * \param[in] conf the configuration
 * \return an error status
 */
//PDI_status_t PDI_EXPORT PDI_init(PC_tree_t conf,const char* configfile, int *is_client,MPI_Comm *damaris_communicator);
PDI_status_t PDI_EXPORT PDI_init(PC_tree_t conf,const char* configfile, int *is_client);


/** Finalizes PDI
 * \return an error status
 */
PDI_status_t PDI_EXPORT PDI_finalize(int);

//PDI_status_t PDI_EXPORT PDI_finalize_Damaris(int *is_client);

/** Checks PDI API version
 *
 * \param[out] provided version if non-null it is filled with the provided API version
 * \param[in] expected if non-zero the expected API version
 * \return an error status if the expected version is incompatible with the
 * provided one
 */
PDI_status_t PDI_EXPORT PDI_version(unsigned long* provided, unsigned long expected);

/// \}

/** \addtogroup annotation
 * \{
 */

/**
 * Access directions
 */
typedef enum PDI_inout_e {
	/// No data transfert
	PDI_NONE = 0,
	/// data tranfer from PDI to the main code
	PDI_IN = 1,
	/// data transfer from the main code to PDI
	PDI_OUT = 2,
	/// data transfer in both direction
	PDI_INOUT = 3

} PDI_inout_t;

/** Shares some data with PDI. The user code should not modify it before
 * a call to either PDI_release or PDI_reclaim.
 * \param[in] name the data name
 * \param[in,out] data the accessed data
 * \param[in] access whether the data can be accessed for read or write
 *                   by PDI
 * \return an error status
 * \pre the user code owns the data buffer
 * \post ownership of the data buffer is shared between PDI and the user code
 *
 * the access parameter is a binary OR of PDI_IN & PDI_OUT.
 * * PDI_IN means PDI can set the buffer content
 * * PDI_OUT means the buffer contains data that can be accessed by PDI
 */
PDI_status_t PDI_EXPORT PDI_share(const char* name, void* data, PDI_inout_t access);

/** Requests for PDI to access a data buffer.
 * \param[in] name the data name
 * \param[in,out] buffer a pointer to the accessed data buffer
 * \param[in] inout the access properties (PDI_IN, PDI_OUT, PDI_INOUT)
 * \return an error status
 * \pre PDI owns the data buffer
 * \post ownership of the data buffer is shared between PDI and the user code
 */
PDI_status_t PDI_EXPORT PDI_access(const char* name, void** buffer, PDI_inout_t inout);

/** Releases ownership of a data shared with PDI. PDI is then responsible to
 * free the associated memory whenever necessary.
 * \param[in] name name of the data to release
 * \return an error status
 * \pre ownership of the data buffer is shared between PDI and the user code
 * \pre PDI owns the data buffer
 */
PDI_status_t PDI_EXPORT PDI_release(const char* name);

/** Reclaims ownership of a data buffer shared with PDI. PDI does not manage
 * the buffer memory anymore.
 * \param[in] name name of the data to reclaim
 * \return an error status
 * \pre ownership of the data buffer is shared between PDI and the user code
 * \post the user code owns the data buffer
 */
PDI_status_t PDI_EXPORT PDI_reclaim(const char* name);

/** Triggers a PDI "event"
 * \param[in] event the event name
 * \return an error status
 */
PDI_status_t PDI_EXPORT PDI_event(const char* event);

/** Shortly exposes some data to PDI. Equivalent to PDI_share + PDI_reclaim.
 * \param[in] name the data name
 * \param[in] data the exposed data
 * \param[in] access whether the data can be accessed for read or write
 *                   by PDI
 * \return an error status
 */
PDI_status_t PDI_EXPORT PDI_expose(const char* name, void* data, PDI_inout_t access);

/** Performs multiple exposes at once. All the data is shared in order they were specified
 *  and reclaimed in reversed order after an event is triggered.
 *
 *  NULL argument indicates an end of the list.
 *
 * \param[in] event_name the name of the event that will be triggered when
 *                       all data become available
 * \param[in] name the data name
 * \param[in] data the exposed data
 * \param[in] access whether the data can be accessed for read or write by PDI
 * \param[in] ... (additional arguments) additional list of data to expose,
 *                each should contain name, data and access, NULL argument
 *                inidactes an end of the list.
 * \return an error status
 */
PDI_status_t PDI_EXPORT PDI_multi_expose(const char* event_name, const char* name, void* data, PDI_inout_t access, ...);

#ifdef PDI_WITH_DEPRECATED

/** Begin a transaction in which all PDI_expose calls are grouped.
 *
 * This requires a call to PDI_transaction_end to close the transaction.
 *
 * \deprecated the transaction part of the API is deprecated, the
 * PDI_multi_expose function should be used instead.
 *
 * \see PDI_expose the function used to expose data inside the transaction
 * \see PDI_transaction_end the function used to end the transaction
 *
 * \param[in] name the name of the transaction (an event thus named will be
 *                 triggered when all data become available)
 * \return an error status
 */
PDI_status_t PDI_DEPRECATED_EXPORT PDI_transaction_begin(const char* name);

/** Ends the previously opened transaction.
 *
 * \deprecated the transaction part of the API is deprecated, the
 * PDI_multi_expose function should be used instead.
 *
 * \see PDI_transaction_begin the function used to start the transaction
 * \see PDI_expose the function used to expose data inside the transaction
 *
 * \return an error status
 */
PDI_status_t PDI_DEPRECATED_EXPORT PDI_transaction_end(void);

#endif // PDI_WITH_DEPRECATED

/// \}

#ifdef __cplusplus
} // extern C
#endif


#endif // PDI_H_
