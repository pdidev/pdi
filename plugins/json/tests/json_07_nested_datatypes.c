/*******************************************************************************
 * Copyright (C) 2023-2024 Commissariat a l'energie atomique et aux energies alternatives (CEA)
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * * Redistributions of source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer.
 * * Redistributions in binary form must reproduce the above copyright
 *   notice, this list of conditions and the following disclaimer in the
 *   documentation and/or other materials provided with the distribution.
 * * Neither the name of CEA nor the names of its contributors may be used to
 *   endorse or promote products derived from this software without specific
 *   prior written permission.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 ******************************************************************************/

#include <paraconf.h>
#include <stdint.h>
#include <stdio.h>
#include <unistd.h>
#include <pdi.h>

const char* CONFIG_YAML
	= "pdi:                                                                 \n"
	  "  data:                                                              \n"
	  "    var_array_record_array_array:                                    \n"
	  "      type: array                                                    \n"
	  "      size: 2                                                        \n"
	  "      subtype:                                               		\n"
	  "        type: struct                                          		\n"
	  "        members:                                           			\n"
	  "          - var_int: int                                          	\n"
	  "          - var_double: double                                       \n"
	  "          - var_array_of_array:                                      \n"
	  "              type: array                                          	\n"
	  "              size: 3                                          		\n"
	  "              subtype:                                          		\n"
	  "                type: array                                          \n"
	  "                size: 4                                          	\n"
	  "                subtype: int                                         \n"
	  "    var_tuple_array_record:                                          \n"
	  "      type: tuple                                        			\n"
	  "      elements:                                          			\n"
	  "        - elem_int:                                          		\n"
	  "          type: int                                          		\n"
	  "        - elem_double:                                          		\n"
	  "          type: double                                          		\n"
	  "        - elem_array:                                          		\n"
	  "          type: array                                          		\n"
	  "          size: 2                                  					\n"
	  "          subtype:                                           		\n"
	  "            type: struct                                          	\n"
	  "            members:                                          		\n"
	  "              - var_int: int                                         \n"
	  "              - var_double: double                                   \n"
	  "  plugins:                                                           \n"
	  "    json:                                                            \n"
	  "      - file: json_07_nested_datatypes.json							\n"
	  "        write:                                           			\n"
	  "          - var_array_record_array_array                             \n"
	  "          - var_tuple_array_record                         			\n";

int main(void)
{
	PC_tree_t conf = PC_parse_string(CONFIG_YAML);
	PDI_init(PC_get(conf, ".pdi"));

	struct record_w_array_of_array {
		int _var_int;
		double _var_double;
		int _var_array[3][4];
	};

	struct record_w_array_of_array _var_raa_1 = {3, 33.3, {{311, 312, 313, 314}, {321, 322, 323, 324}, {331, 332, 333, 334}}};
	struct record_w_array_of_array _var_raa_2 = {4, 44.4, {{411, 412, 413, 414}, {421, 422, 423, 424}, {431, 432, 433, 434}}};

	struct record_w_array_of_array _var_mix[2] = {_var_raa_1, _var_raa_2};

	PDI_expose("var_array_record_array_array", _var_mix, PDI_OUT);

	struct sub_struct {
		int sub_int;
		double sub_double;
	};

	struct tuple_t {
		int elem_int;
		double elem_double;
		struct sub_struct elem_array[2];
	};
	struct tuple_t _var_tuple = {11, 12.3, {{22, 22.2}, {33, 33, 3}}};
	PDI_expose("var_tuple_array_record", &_var_tuple, PDI_OUT);

	PDI_finalize();
	return 0;
}
