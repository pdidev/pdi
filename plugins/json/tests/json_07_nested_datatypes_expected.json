[
{
    "var_array_record_array_array": [
        {
            "var_array_of_array": [
                [
                    311,
                    312,
                    313,
                    314
                ],
                [
                    321,
                    322,
                    323,
                    324
                ],
                [
                    331,
                    332,
                    333,
                    334
                ]
            ],
            "var_double": 33.3,
            "var_int": 3
        },
        {
            "var_array_of_array": [
                [
                    411,
                    412,
                    413,
                    414
                ],
                [
                    421,
                    422,
                    423,
                    424
                ],
                [
                    431,
                    432,
                    433,
                    434
                ]
            ],
            "var_double": 44.4,
            "var_int": 4
        }
    ]
},
{
    "var_tuple_array_record": [
        11,
        12.3,
        [
            {
                "var_double": 22.2,
                "var_int": 22
            },
            {
                "var_double": 33.0,
                "var_int": 33
            }
        ]
    ]
}
]